File: apps/cli/src/runtime/paths.ts
import { existsSync, mkdirSync, chmodSync, unlinkSync } from "node:fs"
import { homedir, tmpdir } from "node:os"
import { dirname, resolve } from "node:path"

export function runtimeDir(env: NodeJS.ProcessEnv = process.env): string {
  const root = env.XDG_RUNTIME_DIR ? `${env.XDG_RUNTIME_DIR}/local-ai-tools` : `${tmpdir()}/local-ai-tools-${process.getuid?.() ?? "user"}`
  mkdirSync(root, { recursive: true })
  try {
    chmodSync(root, 0o700)
  } catch {
    // Best-effort for non-POSIX filesystems.
  }
  return root
}

export function socketPath(explicit?: string | null, env: NodeJS.ProcessEnv = process.env): string {
  return explicit || env.LOCAL_AI_TOOLS_SOCKET || `${runtimeDir(env)}/local-ai-tools.sock`
}

export function pidPath(env: NodeJS.ProcessEnv = process.env): string {
  return `${runtimeDir(env)}/local-ai-tools.pid`
}

export function logPath(env: NodeJS.ProcessEnv = process.env): string {
  return `${runtimeDir(env)}/daemon.log`
}

export function sidecarPath(outputPath: string): string {
  return outputPath.replace(/\.[^/.]+$/, "") + ".json"
}

export function profilePath(outputPath: string): string {
  return outputPath.replace(/\.[^/.]+$/, "") + ".profile.json"
}

export function absolutePath(path: string): string {
  if (path.startsWith("~")) {
    return resolve(homedir(), path.slice(1))
  }
  return resolve(path)
}

export function ensureParent(path: string): void {
  mkdirSync(dirname(path), { recursive: true })
}

export function removeIfExists(path: string): void {
  if (existsSync(path)) {
    unlinkSync(path)
  }
}

File: python/src/local_ai_tools/paths.py
from __future__ import annotations

import os
from pathlib import Path


def runtime_dir() -> Path:
    base = os.environ.get("XDG_RUNTIME_DIR")
    if base:
        root = Path(base) / "local-ai-tools"
    else:
        root = Path("/tmp") / f"local-ai-tools-{os.getuid()}"
    root.mkdir(parents=True, exist_ok=True)
    root.chmod(0o700)
    return root


def socket_path(explicit: str | None = None) -> Path:
    value = explicit or os.environ.get("LOCAL_AI_TOOLS_SOCKET")
    return Path(value) if value else runtime_dir() / "local-ai-tools.sock"


def pid_path() -> Path:
    return runtime_dir() / "local-ai-tools.pid"


def log_path() -> Path:
    return runtime_dir() / "daemon.log"


def sidecar_path(output_path: Path) -> Path:
    return output_path.with_suffix(".json")


def stale_pid() -> int | None:
    path = pid_path()
    if not path.exists():
        return None
    try:
        return int(path.read_text().strip())
    except Exception:
        return None
